### ISSUE TITLE
`nix flake update --single <name>` is more intuitive than `nix flake lock --update`

### ISSUE THREAD

**Issue by colemickens** [NixOS/nix#5110]
Why is `nix flake update` a think instead of just `nix flake lock --update-all`?

Or why is `nix flake lock --update-input` not `nix flake update --single <name>`.

Or is there some other reasoning for the current grouping? I don't get it, I have to look it up every time. It just feels inconistent

**Comment by kubukoz** [comment]
I strongly agree with OP, if Flakes are supposed to make Nix more approachable we should aim to minimize confusion and optimize the UX for discoverability.

**Comment by Kha** [comment]
Or even
```
nix flake update [option...] flake-url [input1 [input2 ...]]
```

**Comment by zyansheep** [comment]
might also be nice to have a `nix flake update --exclude [input1]`

**Comment by brendanzab** [comment]
As a Nix beginner who was just wondering why `nix flake update` didn't let me upgrade specific inputs individually (and had to find out about `nix flake lock` from the forums, I can confirm I found it to be a frustrating papercut!

**Comment by shinzui** [comment]
I have to google this command every time I want to update a single input.

**Comment by erikarvstedt** [comment]
This really needs an UX improvment.
I consider myself a fairly seasoned nix user and I recently wanted to update a single input. This is my story.
- I tried `nix flake update --update-input <input>`. Didn't work.
- I then read through `nix help flake` and the flake-related section of `nix help build`.
- Finally, I settled for `nix eval .#randomInput --raw --apply '_: ""' --update-input nixpkgs-unstable.` 
  I'd never have associated the `lock` verb with updating a flake.

I'd recommend the following:
- Support the common flake arg `--update-input` in `nix flake update`.
- Remove `nix flake lock --recreate-lock-file`.
  What's the difference between this and `nix flake update`?
  If there's a difference, support `--recreate-lock-file` in `nix flake update`.
- Delete subcommand `lock`, replace it with `nix update --add-missing`.

**Comment by TLATER** [comment]
I think there's plenty of agreement here, it'll take a PR to get things moving :)

**Comment by edolstra** [comment]
> It just feels inconistent

It isn't.

* `nix flake update`: Recreates the entire lock file, thus updating everything.
* `nix flake lock`: Add missing locks, optionally updating some locks if `--update-input` is given. This is same behaviour as commands like `nix build`.

See also the discussion in #3781.

**Comment by kubukoz** [comment]
At the very least it's difficult to find 😅

**Comment by robx** [comment]
I just ran into this myself. I saw that `nix flake update` updated my inputs, and it felt very natural to assume I'd be able to restrict that invocation to a single input. In fact, my working solution for the past week was

```
nix flake update; git add -p; git commit -m "partial flake update"; git reset --hard
```

until going off on another lengthy search today and ending up at this issue.

I'd suggest it might useful to override any technical consistency of the current approach in favour of a more intuitive UX?

**Comment by kubukoz** [comment]
I think the following would be more intuitive:

- `nix flake update` - the main way of updating anything in the lock file. Optional argument (named or positional, up to bikeshedding) for updating a specific input
- `nix flake lock` - locks the current inputs (all of them) or one of them if `--update-input` is given. Most likely you'd only ever use this once or never (relying on `nix build` and others to perform the locking)

**Comment by dschrempf** [comment]
I understand both sides but think the status quo is bad.

I think there are two solutions:
1. Either, we remove `nix flake update` completely, and use something like `nix flake lock --update-all`.
2. Or we allow `nix flake update` to act on one input, for example, like so: `nix flake update --only bla`.

**Comment by thufschmitt** [comment]
We discussed that a bit during the last [UX meeting](https://hackmd.io/U21zZ1bwQiW-EnORBaNU1Q?both#nix-flake-update-vs-nix-flake-lock---update-input-foo).

The broad outcome was that:
* Both commands make sense individually 
    * `lock` as a low-level “just do stuff on the lockfile”
    * `update` as the very handy “update everything”
* `update` is morally an alias for `lock --recreate-lockfile`

One suggestion to reduce the confusion was to simply make the fact that `update` is merely a shorthand alias explicit in the documentation (and maybe give `--recreate-lockfile` a better name as it's not entirely obvious what it does)

**Comment by ncfavier** [comment]
This doesn't seem inconsistent with making `update arg1 arg2` an alias for `lock --update-input arg1 --update-input arg2`.

**Comment by thufschmitt** [comment]
> This doesn't seem inconsistent with making `update arg1 arg2` an alias for `lock --update-input arg1 --update-input arg2`.

It's not indeed. However doing so would be inconsistent with the fact that

**Comment by kubukoz** [comment]
Honestly, I doubt beginner users will ever want to update any other flake than `.`.

What's the usecase for `nix flake update github:nixos/nixpkgs`?

Beginners want the ability to easily update:
- all the dependencies
- one particular dependency

I don't think they're gonna want to dig into lockfile-related `lock` commands to find out how to do that. Just `nix update` / `nix update arg1` should be enough, really.

**Comment by thufschmitt** [comment]
Backwards consistency isn't really the matter here (as you say, it's experimental, we could break things).

I agree that taken individually, this command could make more sense in the form `nix flake update input1 input2 ...`. But that would play at the detriment of global coherence: All the commands that can operate on a flake (so nearly all the commands) follow the same convention `nix <command> [flake-uri] [args]`. Breaking this convention would be annoying.

If anything I'd rather remove `update` altogether and recommend something like `lock --update-inputs` as a replacement (and keep `lock --update-input foo --update-input bar`). Or maybe have `lock --update-inputs foo bar` which although a bit longer would have the advantage of preserving consistency and allow a direct list of inputs to update at the same time.

**Comment by thufschmitt** [comment]
> What's the usecase for nix flake update github:nixos/nixpkgs?

That one obviously doesn't make any sense, but `nix flake update /some/directory` does

**Comment by kubukoz** [comment]
I don't think `lock --update-input` helps simplify the UX. How about

- `nix flake update-all` - does what the current `update` does
- `nix flake update` - does what `lock --update-input` does

**Comment by thufschmitt** [comment]
> nix flake update - does what lock --update-input does

Do you mean that you'd call it as `nix flake update input1`?

**Comment by kubukoz** [comment]
> > nix flake update - does what lock --update-input does
> 
> Do you mean that you'd call it as `nix flake update input1`?

@thufschmitt that's right. I see how It's not going to be consistent with the way of passing the flake reference as a positional argument, but sticking to that for the sake of consistency would make for a less intuitive UX for users.

In my view `nix flake update input1` makes sense because positional arguments feel more "important" than named parameters, and the main parameter of "I want to update a single input" is the input you want to update.

**Comment by erikarvstedt** [comment]
`nix flake` currently supports 3 operations for operating on lockfiles, spread over 2 commands.
To improve UX, these should be bundled into one command.

### Lockfile operations

1. Update all inputs / recreate the whole lock file
`nix flake update`
`nix flake lock --recreate-lock-file`

2. Update missing inputs
`nix flake lock`

3. Update a specific input
`nix flake lock --update-input <input>`

### Why the current UX is bad

The split into two commands is confusing. Bundling all lockfile operations into one command would lead to a more intuitive and discoverable interface.

**Example user story**: A user wants to update a specific input of a flake.
- The user runs `nix flake --help` and is currently shown the following subcommand listing, at the top of the help output:
  ```
  nix flake lock - create missing lock file entries
  ...
  nix flake update - update flake lock file
  ```
- The user recognizes `nix flake update` as the relevant command and runs `nix flake update --help`.
- The help shows no info relevant to the task (because only `nix flake lock` supports updating single inputs), the user is stuck.

### Solution: Single command
`update` is the relevant verb for most lockfile operations, so `nix flake update` is the overall best fit for the command name.

Suggested output of `nix flake --help`:
```
nix flake update - update or create flake lock file
```

Suggested design of `nix flake update`:
```bash
nix flake update
nix flake update --update-input <input>
nix flake update --add-missing # replaces `nix flake lock`, which is seldomly called on its own
```

**Comment by edolstra** [comment]
> nix flake update --add-missing # replaces `nix flake lock`, which is seldomly called on its own

I don't think this is true. It's pretty common to add a new input, where you want to update the lockfile to reflect the new input without upgrading every other input. 

I think `nix flake update` should be renamed to `nix flake upgrade` to better reflect what it does (namely upgrade all inputs to the latest version). That avoids the ambiguity of the word "update".

**Comment by erikarvstedt** [comment]
Nix really should fix the `update single input` user story which is catastrophically bad right now.
I'm convinced that a single command that makes the default case simple is by far the most user-friendly design.
Maybe we should hold a vote or do user testing on this issue.

`--add-missing` is no direct negation, it just modifies which inputs are updated.

**Comment by NobbZ** [comment]
> I don't think this is true. It's pretty common to add a new input, where you want to update the lockfile to reflect the new input without upgrading every other input

A `nix flake lock` is indeed rarely necessary, as one usually adds some output using that input and one does `nix build .#thatOutput`, in which case `nix` already implicitely does whatever `nix flake lock` does.

If I do not want this behaviour I have to explicitely opt out using `--no-write-lockfile`, and I have to use that flag in many other cases when nix considers a lock file "stale".

**Comment by LightAndLight** [comment]
> @edolstra:
>
> I do think nix flake update should be renamed to nix flake upgrade to better reflect what it does (namely upgrade all inputs to the latest version). That avoids the ambiguity of the word "update".

Why have two separate commands for lockfile manipulation? Given  that `nix flake lock` has quite a number of features, I suggest that `nix flake update` should be replaced by something like `nix flake lock --update-all`.

**Comment by 9999years** [comment]
I think everyone has been very clear that the current experience is bad, and @edolstra has been clear that he doesn't see a problem or care. Not really sure where to go from this, especially considering how detailed the user stories and feedback written in this issue are.

**Comment by thufschmitt** [comment]
@9999years I don't think anyone denies that there's an issue here. But I don't think we have a clear-cut solution either. 
Everyone (including myself) seems to have their own idea of how this should behave. I, for one, think they are mostly all better than the current status quo, but none seems to stand-out enough to warrant the change. I wouldn't want to do a breaking change now, only to realize in a couple of months that the change wasn't really as good as what it should have been and have to re-break stuff. Maybe it's Buridan's donkey striking again, though, and we should just pick one suggestion at random 🤷‍♂️

**Comment by 9999years** [comment]
> none seems to stand-out enough to warrant the change. I wouldn't want to do a breaking change now, only to realize in a couple of months that the change wasn't really as good as what it should have been and have to re-break stuff

What's the point of `nix-command` being an experimental feature for years if we're not willing to make breaking changes? We should be experimenting with our experimental feature or it should be stabilized.

**Comment by sersorrel** [comment]
As pointed out [a year and two days ago](https://github.com/NixOS/nix/issues/5110#issuecomment-973096650), `nix flake update [option...] flake-url [input1 [input2 ...]]` seems like an unambiguous improvement over the status quo: `nix flake update . nixpkgs home-manager` is dramatically nicer to type than `nix flake lock --update-input nixpkgs --update-input home-manager`, and wouldn't be a breaking change (currently the syntax is `nix flake update [option...] flake-url`, and any additional arguments currently cause the command to print "error: unexpected argument" and exit).

Is anyone willing to implement this?

**Comment by patricksjackson** [comment]
> Is anyone willing to implement this?

My guess is that the implementation is (relatively) trivial, and judging by this thread there is no lack of people who'd be willing to do this.

If I understand correctly, **this issue is blocked waiting on the UX team to make a final decision on the UX.**

In [the linked notes from the UX meeting](https://hackmd.io/U21zZ1bwQiW-EnORBaNU1Q?both#nix-flake-update-vs-nix-flake-lock---update-input-foo) it's unclear what code changes (if any) they decided to take for this, and I don't see a linked work item they made to update the docs.

**Comment by fricklerhandwerk** [comment]
Discussed in Nix team meeting 2022-11-28:

- @fricklerhandwerk: what speaks against using `nix flake update <name>`?
    - @thufschmitt: consistency - the last positional argument is always a flake reference. there are multiple options
        - `nix flake update <flakeref> [<name>]`
        - remove `nix flake update` and use `nix flake lock` for everything
        - leave as it is and state in the manual that the two are aliases of each other (prior decision by now-defunct UX team)
- @tomberek: does anyone have a desire for an outcome here?
    - @thufschmitt: yes, let's ask the high-level question
        - @Ericson2314: the issue is that we have `nix flake <noun>` and `nix flake <verb>`
           - Shaky foundation upon which there probaly *isn't* a good small option, hence the back-and-forth in the command.
           - proposal: write down a table of all commands and see how to reorganise them consistently
- @fricklerhandwerk: proposal: get Elton's UX expertise (see [UX workshop 2022-11-26](https://www.meetup.com/nixos-user-group-zurich/events/289378125/))
    - this will take some time, so this particular issue won't be solved in an hour, but if we spend 4-6 hours in total we could address the whole class of this issues at once
  - agreement. decision:
      - @fricklerhandwerk will ask if Elton or one of his colleagues has time
      - schedule time together to introduce the broader issues and sort them out systematically as in the recent UX workshop
      - collect related issues to address in bulk
        - e.g. maybe https://github.com/NixOS/nix/issues/7261, though it is more 

@patricksjackson the UX team has never been made official and is now disbanded in favor of the Nix team, which is essentially comprised of the same people.

**Comment by patricksjackson** [comment]
> this will take some time, so this particular issue won't be solved in an hour, but if we spend 4-6 hours in total we could address the whole class of this issues at once

Excellent plan :+1:

**Comment by chreekat** [comment]
man pages for the two commands state:

"nix flake lock - create missing lock file entries". Nope, that can't be what I want, the lock file already has the entry.

"nix flake update - update flake lock file". Yes, that must be what I want. I want to update the lock file with my singular dependency update.


If you want to do some UX research, you can look at all the people in this thread who expected to be able to find a way to update a single input with `nix flake update`. And add me to the list. :) I would have never found it in `nix flake lock` if Google hadn't luckily directed me there.

It has the word 'update' right in the name. I don't want to 'lock' something. It's already locked. I want to 'update' it. :)

**Comment by chreekat** [comment]
Sorry for the snark in the last message. If the status quo is considered correct, then perhaps the summaries for the commands should be changed so they do not give people the wrong intuition about how they should be used?

**Comment by chreekat** [comment]
Now I see that `--update-input` is considered a "common flake-related option", so it can be used when running many different kinds of flake commands. That is confusing. What makes "creating missing lock file entries" the subcommand most suitable for updating existing entries anyway?

**Comment by zimbatm** [comment]
I don't know of any other package manager that makes the distinction between lock creation and update. And then moves some of the update into the lock creation. This is so obviously wrong from a UX standpoint, I don't understand why so much discussion needed to happen here.

It should be `nix flake update [<flake-uri>] [--input <input-name>]` and deprecate `nix flake lock` entirely.

Just take Rust as an example: https://doc.rust-lang.org/cargo/commands/cargo-update.html

**Comment by thufschmitt** [comment]
>  I don't understand why so much discussion needed to happen here.

I think everyone agrees that the current state is bad, but

1. No one agrees as to what the semantics should be 
2. Nobody (unless I missed it, in which case my sincere apologies) opened a pull-request to change the behavior.

I personally agree with 

> It should be nix flake update [<flake-uri>] [--input <input-name>] and deprecate nix flake lock entirely.

and would welcome an implementation of it :)

**Comment by joneshf** [comment]
> I do think `nix flake update` should be renamed to `nix flake upgrade` to better reflect what it does (namely upgrade all inputs to the latest version). That avoids the ambiguity of the word "update".

Just an anecdote from one user of nix that has used a slew of other package managers. The difference between update and upgrade probably makes sense from an implementer's perspective. You should do make the change if it helps you internally maintain nix. But from a user's perspective, it really doesn't translate. Most people see update and upgrade as synonyms, so it's not going to change things from a user's perspective in an noticeable way.

**Comment by iFreilicht** [comment]
I see this is a important issue that lots of people are interested in, and for good reason. I have capacity right now and would like my hours to be high-impact, so I'm willing to implement this. Of course, we need to reach a conclusion first.

# My proposal

Keep two commands. `nix flake lock` makes sense for the initial operation of creating a lockfile. It's easy to remember and clear in what it does, it "*lock*s the flake". `nix flake update` allows for more control and is purpose-made to "*update* the inputs". This sentiment was repeatedly mentioned above, so let's try to lean into it.

## `update`

`nix flake update` [*options...*] *flake-ref* [*inputs...*] - update flake inputs

### Examples

* Get explanation of how `update` works
```console
$ nix flake update
error: Too few arguments. Please explicitly specify the flake and inputs to update.
hint: If you just want to update all inputs of the flake in the current directory, you can run
      nix flake update --all
```

* Update all currently locked inputs (also removes missing ones)
```console
$ nix flake update --all
```

* Update all inputs of any flake (after new inputs were added to `flake.nix` that were not locked yet)
```console
$ nix flake update --all /path/to/that/flake
warning: New inputs were added that are not locked yet.
hint: To add all new inputs to the lockfile, run
      nix flake lock /path/to/that/flake
```

* Update a single input of the current flake
```console
$ nix flake update . nixpkgs
```

* Update multiple inputs of any flake
```console
$ nix flake update /path/to/that/flake nixpkgs home-manager flake-utils
```

* Error message when only supplying *flake-ref* to avoid accidental "success"
```console
$ nix flake update nixpkgs
error: No inputs provided. Please specify which inputs to update or use '--all' to update all inputs.
hint: Maybe you wanted to update the 'nixpkgs' input of the current flake? If so, run
      nix flake update . nixpkgs
```

* Only remove inputs from the lock that are missing from `flake.nix`
```console
$ nix flake update --only-remove-missing
```

## `lock`

`nix flake lock` [*flake-ref*] - lock flake inputs

### Examples

* Create lock for current flake or add missing inputs to current flake's lock
```console
$  nix flake lock
```

* Create lock or add missing inputs for any flake (not sure how useful that is, but it makes sense for consistency)
```console
$ nix flake lock /path/to/that/flake
```

## Reasoning

I think this gives both commands clear and separated purpose. `lock` is simple and makes it clear that no existing locks will be modified. `update` will not lock new inputs, but this may be surprising, so it warns you if you run it after adding new inputs to `flake.nix` that you didn't lock yet.

It may be controversial to make `nix flake update` fail without arguments, but I feel this is necessary to create an interface that protects the user from accidental actions. It's not clear that the default should be to update all packages, and reading this thread it seems I'm not the only one in that camp.

It may also be controversial to allow additional positional arguments after *flake-ref*. My only argument here is that it's easier to type than `--inputs` or `--only`, which also creates questions around how to order the arguments, and at least to me it *feels* pretty unambiguous in meaning.

There's of course potential to make `update` more powerful; it could also create the lockfile by default, and with an `--add-missing` flag, it would basically replace `lock`, though I think in that case it would still be fine to leave `lock` as an alias for that, just for the brevity of it.

About the `hint:` lines; I know there's no such mechanism yet, this just my idealized version of teaching the user directly in the interface, plus easy command copying from #8729. Adding something like `Error.addHint` might prove very useful for lots of tools in nix, though.

I do like @YorikSar's idea of re-thinking the structure of `nix flake` and discoverability general. I believe the current verbs were chosen well, and even deeper nested commands are annoying, so I vouch for `lock` and `update` to stay, though they could of course become `nix flake inputs lock` and `nix flake inputs update` without changing my proposal above. This feels out-of-scope for this specific feature request, though.

**Comment by 9999years** [comment]
@iFreilicht That mostly sounds good, but I think `nix flake update` should update all the inputs. I think if we require `--all` we'll have lots of people try `nix flake update` and get annoyed that it doesn't do anything by itself. (I know I'd certainly reach for `nix flake update` before `nix flake update --all`.)

**Comment by thufschmitt** [comment]
Thanks for the proposal @iFreilicht !

It feels pretty good to me. I'll try and get some feedback from the rest of @NixOS/nix-team (although I'm going on vacation this evening and until eom, so I won't be able to push for this a lot)

**Comment by chreekat** [comment]
Thanks @iFreilicht !

My 2 cents:

* `nix flake update [--flake <flake-url>] [<input>...]`: update inputs to the flake. With no arguments, defaults to updating all inputs of `./flake.nix`, just like today.
* `nix flake lock [<flake-url>]`: same behavior as today.
* `nix flake lock [--flake <flake-url>]`: same as previous command, but provided to have a UI consistent with `nix flake update`.

I have no opinion on the other use cases, because I haven't used them yet.

**Comment by 9999years** [comment]
One thing I don't want to lose sight of: just about _anything_ is better than the CLI UX as it stands right now, and we should be taking advantage of the fact that flakes/nix3 aren't stabilized yet to do this UX experimentation while we can.

**Comment by iFreilicht** [comment]
> but I think `nix flake update` should update all the inputs

@9999years I'm somewhat cautious to make a potentially destructive action the default. If the current state of the lock is not tracked by version control (i.e. git), updating all inputs is an action that is impossible to reverse. This may happen if:

1. we operate on a `path:` flake. This *may* just [be a workaround](https://discourse.nixos.org/t/help-with-python-flake-build-from-local-source/30140/6?u=ifreilicht), but there's no guarantee
2. the lock hasn't been checked-in yet, maybe it was just created or the user simply forgot

I recognize those are a bit of an edge-case, so I'm happy to optimize for the happy path. As a solution, I propose a **confirmation dialog that only pops up when Nix can't guarantee that the changes are reversible**. Nix would check if the flake is a git flake _and_ if the path `flake.lock` is checked-in and clean; if both of those are the case, we know the change is reversible, and _the dialog is skipped_. If not, it gives you an explanation and something like:

```
$ nix profile update
'nix profile update' updates all inputs by default. [You're operating on a non-git flake/The path 'flake.lock' is dirty], so this is a potentially irreversible action. If you just want to update a few select inputs, run 'nix flake update . <inputs>'.
Do you really want to update all inputs? (Run with '--all' to skip this dialog) (y/N)
```

We already have [two dialogs of this nature](https://github.com/NixOS/nix/blob/master/src/libexpr/flake/config.cc#L62-L67) when evaluating [a flake that contains non-whitelisted keys](https://github.com/NixOS/nix/blob/master/src/libexpr/flake/config.cc#L34) in its [`nixConfig` attribute](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-flake#flake-format).

> do this UX experimentation while we can

Yes indeed. I'll definitely start working on a PR, even without getting the idea approved first. That way interested people can also try it out with a simple `nix run github:iFreilicht/nix-flake-update`. That's even quicker than getting things merged and waiting for community feedback.

> `nix flake update [--flake <flake-url>] [<input>...]`

@chreekat hmm I do see the appeal, but this would be very inconsistent with any other command taking an installable or flake. What do you think of the solution I proposed above? It would retain the current behavior of `nix flake update` in most cases, and `nix flake update . nixpkgs` is only two keystrokes more than your solution.

**Comment by chreekat** [comment]
@iFreilicht your suggestion is an improvement on the status quo, so in that regard I think it's fine.

That said, I think "it's inconsistent with any other command" is kind of an anti-argument, given our regard for the status quo. :sweat_smile: Taken on its own merits, specifying which flake to update seems like a niche use case (I've never used it), so specifying it as a required argument seems un-ergonomic.

Talk is cheap, though, so please do proceed as you see fit, regardless of my opinions.

**Comment by TLATER** [comment]
I agree with @chreekat on the positional argument. I think a `--flake` arg is less confusing.

This is plain unreadable:

```nix
nix flake update /path/to/that/flake nixpkgs home-manager flake-utils
```

I have no idea what this command is doing here. Ok, so I'm updating a flake, and that first thing is a path, seems to be a flake, so given the subcommand it probably updates that. But what the hell are these other args about? Mixing the meaning of positional args always results in confusing CLIs IMO, and looks bad when reading scripts.

**Comment by colemickens** [comment]
re: @iFreilicht sorry if I've missed something catching up, but did you mean `nix profile update` there, really? Or was that maybe a typo/brain-slip? As far as I can tell that's the first mention and the profile commands aren't for working with the `flake.{nix,lock}`.

**Comment by iFreilicht** [comment]
Yeah after starting to look at the codebase and experimenting a little bit, `flake update`  (and `flake lock` for that matter) is a bit of an outlier here. For most commands, it is very useful to be able to pass any flakeref. Put for `update`, not so much, the flake needs to be writable for it to have any effect. You _can_ run `nix flake update nixpkgs` and it _will_ complete, but it's basically a no-op.

So the only scenario is for a different path, not flake-refs outright, and I'd be very surprised if there were so many people doing that, that `nix flake update /path/to/flake` as opposed to just, `cd /path/to/flake && nix flake update` would be an unreasonable burden. The only real reason is explicitly specifying the flake type, like `path:.`, but having a `--flake` arg as an escape hatch for that seems reasonable. It's only necessary in a few select cases.

**Comment by colemickens** [comment]
Sorry to wade back in after being absent from this thread, but I have some strong feelings.

1. I think abusing positional arguments is a very unfortunate idea and will be regretted quickly if chosen. I do not like the proposed options that make it very unclear what the args are at a quick glance. I can imagine the exact bugs it will cause right now.  Rather than the design including special handling to pre-empt user misuse, we should strive for intuitive, explicit CLI UX.

2. Regarding operating on flakes outside of PWD.... I think this should be solved like Git. Which also has the nice benefit of making pt1 easier to avoid.

```shell
cd /this/is/not/a/git/worktree
git clone https://some/flake ~/myflake
git -C ~/myflake rebase -i
nix flake -C ~/myflake lock --recreate-lock-file
nix flake -C ~/myflake lock --override-input nixpkgs ~/nixpkgs_fork
nix flake -C ~/myflake update # doesn't do anything, spits out a warning and usage info
nix flake -C ~/myflake update --all # cmd2
nix flake -C ~/myflake update --input nixpkgs-wayland
nix flake -C ~/myflake update --input nixpkgs --input nixpkgs-wayland
```

**Comment by 9999years** [comment]
@colemickens Please show me an example from the dozens of existing dependency management tools where `the-tool update` prints out a warning and usage info instead of, y'know, running the update.

**Comment by iFreilicht** [comment]
> I think abusing positional arguments is a very unfortunate idea and will be regretted quickly if chosen. 

Just to be clear, are you talking about mixing the types of positional arguments like I proposed?

```
$ nix flake update . nixpkgs nixpkgs-wayland
```

Or are you talking about using positional arguments in general, like @TLATER and @chreekat proposed?

```
$ nix flake update nixpkgs nixpkgs-wayland
```

The former I agree with, it's not a great idea, it's confusing an error-prone. But I'm really having a hard time seeing the issue with the latter one.

> I can imagine the exact bugs it will cause right now.

What would those be, roughly?

> we should strive for intuitive, explicit CLI UX

Sure, but we seem to disagree what intuitive means. Intuition is built from our individual experiences, and so what we deem intuitive differs. I of course appreciate you voicing your opinion, but the general sentiment of this thread seems to be that the most common usecase should be the simplest and most obvious one, and so what's intuitive in this context is a democratic decision.

> `nix flake -C ~/myflake lock --recreate-lock-file`

I do like the idea, kind-of, but operating on any flakeref is very useful and common for almost all `nix flake` commands _except_ `update`. So changing the existing UX from `nix flake metadata nixpkgs` to `nix flake -C nixpkgs metadata` seems to be an overall regression. It seems like that's what you're suggesting.

> maybe for fun nix flake update could actually do a dry-run and print a summary of what inputs could be updated, and then encourage the user to `nix flake update --all` or `nix flake update --input [one_from_the_summary]`

I do like that idea, and I will try to implement it for flakes where this action isn't known to be reversible. I'd still argue that offering a simple yes/no dialog is the user friendly and low-friction solution, as opposed to erroring out, but a little encouragement isn't misplaced here.

**Comment by iFreilicht** [comment]
@colemickens Ahh I just saw that you're actually the person who opened this ticket in the first place! After reading through the code for the last hours, let me give you an answer to your initial questions:

> Why is `nix flake update` a think instead of just `nix flake lock --update-all`?
>
> Or why is `nix flake lock --update-input` not `nix flake update --single <name>`.
>
> Or is there some other reasoning for the current grouping? I don't get it, I have to look it up every time. It just feels inconistent

The reason is that Nix actually tries very hard to be 100% consistent to the point of being confusing.

Under many commands in the nix manual, you'll see a section "Common flake-related options:". [Any command that can take a flake as an input can take these](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-shell.html?search=--update-input). I think for something like `--no-use--registries`, or `--override-input`, this makes perfect sense as it changes how a flakeref is resolved. 
But `--update-input` and `--recreate-lock-file` are part of that list, and those actually (try to) mutate the flake you're referring to!

This does enable you to call `nix build --recreate-lock-file` instead of `nix flake update; nix build`, but I can't see how the former is better than the latter. No idea how many people know that this is possible or have even used this feature once.

And these flags can of course also be parsed to `flake update` and `flake lock`. Actually, `nix flake update` right now is **the exact same** as `nix flake lock --recreate-lock-file`, it just sets the `--recreate-lock-file` flag internally and ignores all `--update-input`s.

I'm not entirely sure what the motivation was for this design, I guess to have primitives you could build command invocations from? No matter what, this generalization clearly gets in the way when the _only_ thing you want to do is update inputs.

**Comment by roberth** [comment]
Discussed in meeting:

We've had a discussion about the design. We would love for this UX issue to be made some progress on, and here's our thoughts on it:

Low hanging fruit is adding `nix flake update --update-input` support which is consistent with the existing CLI.

Detailed notes:

- @thufschmitt (through agenda): Old issue that never got solved. We have a sensible proposal and someone willing to implement it. I say go for it!

#### `--all`

- @edolstra: `--all` is similar to https://github.com/NixOS/nix/pull/4809. What's the expectation? I'd expect it to update everything
- @ericson2314: the math thing suggests not to do anything (disjunction identity)
- @roberth: `--all` is more pain than the possible pain someone writing a script might encounter
- Conclusion: avoid `--all`; just update everything

#### `-C` and meaning of positional arguments

- @edolstra, @tomberek: dynamically interpreting positional argument one would be ambiguous
- @roberth: `nix update` instead of `nix flake update` so that we have an excuse to diverge from the "arguments are flakerefs" convention
- @edolstra: Avoid `-C` because it's inconsistent with the rest of the CLI
- Preliminary conclusion: no changes then?
- @tomberek: Can we acknowledge there's a UX issue?
- @tomberek: Switching between `nix flake update` and `nix flake lock` is counterintuitive

#### Locking flags on all flake commands

- @ericson2314: Don't all commands support `--update-input`?
- @edolstra: Yes, except `nix flake update`, because it's equivalent to `--recreate-lock-file`, so `--update-input` would be meaningless
- @ericson2314: Flags should be adjectives/adverbs not verbs
- @edolstra: However users expect `nix flake update` to have a flag to update just the specified inputs (e.g. `--single`)

#### Brevity and lock command expectations

- @roberth: I think this is a (rare-ish) case where the UI brevity is actually really important, so we should consider `nix update` as an alias for the very specific behavior that people want but is not consistent with the otherwise verbose CLI conventions
- @ericson2314: Can we sidestep the issue by reconsidering the basics, verbs, adjectives
  - Also, looking at the implementation I see `removeFlag`
  - @edolstra: Well, some commands don't make sense with certain global Flake flags.
  - @Ericson2314: I don't disagree, but I think this is a problem we should be having -- if we have to filter-out global flags, this is evidence those should have never been global flags to begin with. 
      - @edolstra: This is a side-effect of `nix flake update` being a shorthand for `--recreate-lock-file`. Normally you would never pass `--recreate-lock-file --update-input ...`.
- @edolstra: `--update-input` is a bit weird to have in all commands, but `--override-input` is rather useful
- @edolstra: `--update-input` could be removed from the global flags
- @roberth: `--override-input` is like `--with-overridden-input` but less verbose
  - @ericson2314: +1
  - @ericson2314 & @roberth: conversely the persistent side effect ones (such as `--update-input`) don't make sense as `--with-...`. This is good evidence that they don't belong there! (and just belong on `update`/`lock`).
    - hypothesis: the non-adjective/adverb ones that cannot be rephrased (e.g. with `--with-...`) are the sketchy ones which need to be filtered out today with `removeFlag`!

**Comment by iFreilicht** [comment]
> * avoid `--all`; just update everything

Good, it seems that's in line with the expectations of those who raised their voice so far.

> * Avoid `-C` because it's inconsistent with the rest of the CLI

Agreed.

> * `nix update` instead of `nix flake update` so that we have an excuse to diverge from the "arguments are flakerefs" convention

We _have_ an excuse to diverge from the convention, which is; the UX is bad and there's no good reason to apply the convention in this particular case. Having `nix update` also makes that command seemingly unrelated to flakes when in reality it is *only* for flakes.

> * `--update-input` could be removed from the global flags
>     * if we have to filter-out global flags, this is evidence those should have never been global flags to begin with. 
>     * conversely the persistent side effect ones (such as `--update-input`) don't make sense as `--with-...`. This is good evidence that they don't belong there! (and just belong on `update`/`lock`).

I'm taking the freedom of seeing this as confirmation to my statement above and will remove `--update-input` and `--recreate-lock-file` in my draft PR. I can always revert that change before merging, but I'd like to see how it *feels*.

**Comment by 9999years** [comment]
> Having `nix update` also makes that command seemingly unrelated to flakes when in reality it is _only_ for flakes.

IME this is a big source of confusion for new users who don't understand that everything under `nix` is for flakes (especially for commands like `nix search` which don't have equivalents in the nix2 CLI).

**Comment by Ericson2314** [comment]
@9999years Just so we are claer, it is an explicit goal of https://github.com/NixOS/rfcs/pull/136 that the new CLI *not* be *exclusively* for Flakes, however. (That one is in FCP and close to being merged unless something new comes up.)

**Comment by iFreilicht** [comment]
> that everything under `nix` is for flakes

Though not everything there is _only_ for flakes, but _installables_, which indeed includes flakes, but also files with `-f`/`--file`, store paths, derivations and derivation outputs and bare Nix expressions with `--expr`.

But the `nix flakes` commands _only_ operate on flakes.

**Comment by iFreilicht** [comment]
The PR is ready: #8817 

For any curious users that want to test it, I added a `Testing` section to the PR. Just follow the instructions there.
